West 0.9 docs

mbolivar-nordic opened the pull request:
Documentation for the latest west release.

mbolivar-nordic committed:
doc: west has moved to allowlist/blocklist

The west manifest import feature now uses the terms "allowlist" and
"blocklist" instead of "whitelist" and "blacklist", respectively.

The old terms are still supported by the tool for compatibility, but
the docs for this feature are going to use the new terms exclusively.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: release notes for 0.9.0

The main new features are submodule support and project groups.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west update can create loose commits

Try to spell out the implications of that a bit more clearly.

Suggested-by: Marc Herbert <marc.herbert@intel.com>
Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: add submodules

West 0.9, now with submodules!

This is a new feature which allows you to incorporate a project's Git
submodules into the workspace.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: update schema version

West 0.9 has a new schema version; document it and fix a missed update
from 0.8. Whoops.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: API changes for 0.9.0

These are minimal.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: re-work west update to prep for new features

The west update docs mix together the steps for updating a project
with the set of projects that are being updated. This is inconvenient
now that west 0.9 has a concept of inactive projects.

Split the content into a part that describes *which* projects are
updated, and a separate part that describes how *each* project is
updated.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@kkasperczyk-no please review the submodule portion if you can.

marc-hb commented on doc/guides/west/repo-tool.rst:
```diff
@@ -226,37 +226,118 @@ manifest` file.
+   git submodule update init --rebase --recursive
+
+Otherwise, the project has ``submodules: <list-of-submodules>``. In this
+case, west one of the following in the project repository for each
```
Missing a word or two

mbolivar-nordic replied:
Missing the word "runs". Thanks.

kkasperczyk-no commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,428 @@ processed in this order:
+it's ``false`` or missing, it has no effect.
+
+For example, let's say you have a source code repository ``foo``, which has
+some submodules, and you want ``west update`` to keep them in sync, along with
```
It's quite understandable that this way is used for updating all submodules, but I would just emphasize it by: "... and you want ``west update`` to keep all of them in sync ...".

kkasperczyk-no commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,428 @@ processed in this order:
+   manifest:
+     projects:
+       - name: foo
+         submodules:
```
I would recommend adding one more submodule in order to demonstrate the list having multiple (at least two) elements.
For example:
```
- name: foo
   submodules:
      - name: foo-first-sub
        path: path/of/first_submodule/you/want
      - name: foo-second-sub
        path: path/of/second_submodule/you/want
- name: bar
```

tejlmand replied:
I got to think.

Should we create an enhancement task here ?
What if `foo-first-sub` also contains submodules that I want, but I still don't want to enable all submodules for `foo`.

kkasperczyk-no replied:
The current implementation will update `foo-first-sub` and all of its submodules recursively, so I guess that addresses your question. Another question could be - what if `foo-first-sub` also contains submodules and I don't want to update them, but only the `foo-first-sub`. The second use case is currently not possible to do, as all submodules are updated recursively.

marc-hb replied:
I think someone who has complex submodules need should just disable them in west. Submodules are complex enough, if/when I have to actually deal with submodules the last thing I'd want is an incomplete layer of indirection on top of them. I see this new feature as a basic, simple "download all the pieces" feature for the users who do NOT have to touch submodules and don't even want to know about them.

tejlmand replied:
> The current implementation will update foo-first-sub and all of its submodules recursively, 

Then that is a very important piece of information that is missing.

kkasperczyk-no reviewed (commented):
LGTM, I left just two small remarks.

mbolivar-nordic commented:
> LGTM, I left just two small remarks.

Thanks @kkasperczyk-no ; both comments addressed.

kkasperczyk-no commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,430 @@ processed in this order:
+project's absolute path in the workspace -- to use for the submodule.
+
+For example, let's say you have a source code repository ``foo``, which has
+many submodules, and you want ``west update`` to just keep one of them in sync,
```
After adding second submodule to example I would replace "one" with "two" or "few".

mbolivar-nordic replied:
Oh, nice catch, thanks. I changed to 'some but not all'.

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -219,7 +221,14 @@ The list of project keys and their usage follows. Sometimes we'll refer to the
+  :ref:`west-manifest-import` for details.
+- ``groups``: Optional, a list of groups the project belongs to. See
+  :ref:`west-manifest-groups` for details.
+- ``submodules``: Optional. If present, ``west update`` will also update `Git
```
It's a bool or a list, so this sentence is technically not accurate? i.e. `submodules: false`

mbolivar-nordic replied:
Good catch, thanks. Will change to:

```
You can use this to make ``west update`` also update `Git submodules`_
defined by the project.
```

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -768,12 +778,12 @@ The ``import`` key can also contain a mapping with the following keys:
-Whitelists override blacklists if both are given. For example, if a project is
-blacklisted by path, then whitelisted by name, it will still be imported.
+Allowlists override blocklists if both are given. For example, if a project is
+blocklisted by path, then allowlisted by name, it will still be imported.
```
Do those verbs actually exist? :) not sure if "blocked" and "allowed" would be better here

carlescufi replied:
Same for other instance further down.

mbolivar-nordic replied:
Ah, the perils of search and replace. Thanks.

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -831,16 +841,16 @@ An equivalent manifest in a single file would be:
          url: https://git.example.com/mainline/lib2
 
-If a whitelist had not been used, the ``lib`` project from the mainline
+If a allowlist had not been used, the ``lib`` project from the mainline
```
```suggestion
If an allowlist had not been used, the ``lib`` project from the mainline
```

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,430 @@ processed in this order:
+**********************************
+
+You can use the ``groups`` keys briefly described :ref:`above
+<west-manifest-files>` to put projects into groups. The ``groups`` key can be a
```
```suggestion
<west-manifest-files>` to place projects into groups. The ``groups`` key can be a
```
maybe?

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,430 @@ processed in this order:
+     projects:
+       - name: some-project
+         groups: ...
+     groups: ...
```
Can you move this above `projects:` so that  the order in the file matches the description above?

mbolivar-nordic replied:
I don't follow this one. The description above says:

```
The ``groups`` key can be
a project or top level manifest attribute
```

This ordering matches that description.

carlescufi replied:
Oh, I misread the "top-level attribute" sentence. You are indeed right.

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,430 @@ processed in this order:
+         groups: ...
+     groups: ...
+
+You can allow or block project groups. Projects whose groups are all blocked
```
I wonder if we should use "enable" and "disable" instead of "allow" and "block". To me the former feels truer to the original intent, but this is just a suggestion.

carlescufi replied:
Same for other instances below

mbolivar-nordic replied:
I'm trying to adopt 'allow' and 'block' as consistent replacements for 'white' and 'black' in general here, so I'd prefer to keep as-is if that's OK.

carlescufi replied:
Yes, but isn't enabling or disabling groups conceptually different from allowing or blocking projects? From my perspective it is, because projects are allowed or blocked directly, whereas enabling or disabling groups then has an impact on what projects end up being active or inactive, which is different from them being allowed or blocked.

tejlmand replied:
when using `active` and `inactive`, then I feel `enable` and `disable` are fitting better.
But that's just my feeling as non-native English speaker.

marc-hb replied:
> Yes, but isn't enabling or disabling groups conceptually different from allowing or blocking projects? 

This question confused me a lot because you crossed the adjectives and the names. Did you do that on purpose to illustrate your own confusion?

Here's my non-native speaker perspective too:
- The adjectives active and inactive sound (to me) like "final states", they do not like some indirect desire that gets combined with and can takes precedence over others desires/groups. So it's perfect for projects.
- The verbs allow/block do sound like indirect desires. So they're good for groups.
- At first I found the verbs enable/disable sounded better than allow/block because they don't have any security connotation, so they reflect the intent better. However they feel less "indirect" to me, maybe because it's more common to combine security permissions or because they're very close to the adjective enabled/disabled which sound like final states.

marc-hb replied:
Oops, I just realized "allow/block" is used for groups in group filters but the same terms are use for projects in `import` :-(

> trying to adopt  'allow' and 'block' as consistent replacements for 'white' and 'black' in general here

I think that particular  "migration consistency" should have lower priority than all other "self-consistencies". What is its purpose, make search/replace easier? That would sound like a small and short-term gain.

carlescufi replied:
I am not sure myself. I stand by the comments I've made, and @marc-hb and @tejlmand seem to align, but I can't say that I am convinced. Let me try to summarize so I can try to clarify it for myself and others:

1. Keep as-is:
   - groups can be allowed or blocked
   - projects are active or inactive
   - projects can be placed in an allowlist or blocklist when importing
2. Change to:
   - groups can be enabled or disabled
   - projects are active or inactive
   - projects can be placed in an allowlist or blocklist when importing

I think I prefer 2., because allowing or blocking has a ring to "external". You allow or block something that is defined externally and then you import or deal with. So that is very good for importing, because you are either allowing or blocking a project that is not controlled by you. But groups are described and managed in-manifest, so that is why I think enabling and disabling them makes more sense that allowing them or blocking them.

tejlmand replied:
> * Keep as-is:
>   
>   * groups can be allowed or blocked
>   * projects are active or inactive
>   * projects can be placed in an allowlist or blocklist when importing
> 
> * Change to:
>   
>   * groups can be enabled or disabled
>   * projects are active or inactive
>   * projects can be placed in an allowlist or blocklist when importing

Thanks for this.
Thinking a bit more on this, I start to like 2. better and better, as it removes any ambiguity regarding `allow` / `block`.
`allow` is solely used in relation to import (`allowlist`), so will never be mixed up with groups in any way, as groups are enabled or disabled.

mbolivar-nordic replied:
OK, option 2. it is. I've updated to the documentation to match the code changes in https://github.com/zephyrproject-rtos/west/pull/466.

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,430 @@ processed in this order:
+Project Groups
+==============
+
+Inside ``manifest: projects:``, you can add a project to groups. The ``groups``
```
```suggestion
Inside ``manifest: projects:``, you can add a project to one or more groups. The ``groups``
```

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,430 @@ processed in this order:
+
+The groups named ``groupA`` and ``groupB`` are blocked.
+
+Groups are otherwise allowed by default.
```
I feel that the enabling/disabling of groups (or allowing/blocking in your current terms) should be featured at the beginning of this section, so the hierarchy is respected: first explain what it means to enable or disable groups, then explain what effect this has on projects themselves being active or inactive.

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,430 @@ processed in this order:
+You can think of this as if the ``manifest.groups`` configuration option is
+appended to the ``manifest: groups:`` list from YAML, with "last entry wins"
+semantics.
+
```
Maybe mention here that there's a command-line option to west that has the highest precedence? So that users are aware of it when reading about groups.

mbolivar-nordic replied:
> Maybe mention here that there's a command-line option to west that has the highest precedence? So that users are aware of it when reading about groups.

I will document this in repo-tool.rst with the rest of the ``west update`` details if that's OK.

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,430 @@ processed in this order:
+The ``manifest.groups`` configuration option is set to ``-groupA`` (you can
+ensure this by running ``west config manifest.groups -- -groupA``; the extra
+``--`` is required so the argument parser does not treat ``-groupA`` as a
+command line option ``-g`` with value ``roupA``).
```
```suggestion
command line option ``-g`` with value ``groupA``).
```

mbolivar-nordic replied:
No, this one is correct.

carlescufi replied:
Yep, sorry my mistake.

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,430 @@ processed in this order:
+In this case, both ``groupA`` and ``groupB`` are blocked.
+
+Therefore, projects ``foo`` and ``bar`` are both inactive.
+
```
I don't want to be a pain, but I think we should also include an example using the command-line option to `west update` here.

carlescufi reviewed (changes requested):
Excellent doc as usual. Some comments and suggestions.

mbolivar-nordic committed:
doc: west: add manifest.file config option

I never added documentation for this option; whoops.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@carlescufi thanks for review. I addressed your comments and rebased onto the latest #31201 (in separate force pushes, to ease looking at the changes).

marc-hb commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,433 @@ processed in this order:
+This is the easiest way to use ``submodules``.
+
+If ``submodules`` is ``true`` as a ``projects`` attribute, ``west update`` will
+update the project's Git submodules whenever it updates the project itself. If
```
recursively

marc-hb commented on doc/guides/west/repo-tool.rst:
```diff
@@ -224,39 +224,142 @@ manifest` file.
+   HEADs <https://git-scm.com/docs/git-checkout#_detached_head>`_
+#. If the manifest file specifies a :ref:`submodules
+   <west-manifest-submodules>` key for the project, updates the project's
+   submodules as described below.
```
recursively

marc-hb commented on doc/guides/west/repo-tool.rst:
```diff
@@ -224,39 +224,142 @@ manifest` file.
+remotes even if the revisions appear to be available locally, either use
+``-f always`` or set the ``update.fetch`` :ref:`configuration option
+<west-config>` to ``always``. SHAs may be given as unique prefixes as
+long as they are acceptable to Git.
```
Can you add something like: "as numerous git servers don't allow fetching SHAs directly, west fetches all branches (and tags?) when the revision looks like a SHA and is not found locally. See issue 1234 for future improvements."

mbolivar-nordic replied:
I've added a commit documenting the existence and reservation of `refs/west`, but that's all I'd like to document for now, as just about everything related to this is an implementation detail.

marc-hb replied:
Interesting but that wasn't my point. My main concern is _performance_ misconceptions. `revision:` fields in the manifest are misleading because they look like they let git fetch only what's required - which is indeed possible but only on some servers so west has to be inefficient and slow.

Some background:
https://opensource.googleblog.com/2018/05/introducing-git-protocol-version-2.html
(Note the main optimization described there is way ahead of this)

mbolivar-nordic replied:
I think this is a bit of a digression for this text so I would like to put it in a footnote. LMK if that's no good.

marc-hb replied:
Footnote or anywhere else is fine, this is "only" about documenting some network and storage aspect. Probably west's main performance aspect though.

BTW this would also indirectly explain why updating a `revision: deadbeef` or `revision: cafe5` branch that is not prefixed with `refs/heads/`, can be much slower than updating `revision: java5`, correct?

mbolivar-nordic replied:
> Footnote or anywhere else is fine, this is "only" about documenting some network and storage aspect. Probably west's main performance aspect though.

OK, that's done so I'm going to leave this as 'resolved'. Thanks.

> BTW this would also indirectly explain why updating a `revision: deadbeef` or `revision: cafe5` branch that is not prefixed with `refs/heads/`, can be much slower than updating `revision: java5`, correct?

Yep.

mbolivar-nordic committed:
doc: west reserves refs/west/

These Git refs are used by west as a scratch space. Let that be known,
but don't make any other guarantees.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

tejlmand commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1183,357 @@ processed in this order:
+**********************************
+
+You can use the ``groups`` keys briefly described :ref:`above
+<west-manifest-files>` to place projects into groups. The ``groups`` key can be
+a project or top level manifest attribute, like this:
```
When reading this, a user might think that the `groups` key is is used for same purpose, placing a project in a group, regardless of whether the `groups` key is placed at project or manifest level.

But the fact is, that the `groups` key at project level describes to which groups a project belongs, and at manifest level it specifies what groups to en-/disable.

Could you re-phrase this ?
Sorry for not having a proposal here.

marc-hb replied:
> But the fact is, that the groups key at project level describes to which groups a project belongs, and at manifest level it specifies what groups to en-/disable.

I find it unfortunate that these two different keys have the same name.

mbolivar-nordic replied:
> I find it unfortunate that these two different keys have the same name.

Would you like to propose an alternative? We haven't released this yet. I'm open to suggestions.

tejlmand replied:
Ok, let me give this a try, feel free to improve further:

>  At project level the ``groups`` defines the groups to which the project belong, at the top level manifest attribute the ``groups`` enable or disable groups, like this:

mbolivar-nordic replied:
> Ok, let me give this a try, feel free to improve further:

@tejlmand I think @marc-hb is saying he would like to rename "manifest: groups:" to "manifest: something-different-than-groups:".

tejlmand replied:
ok.
but I still think it would be valuable for the reader if it was highlighted that the two `groups` key are used for different purposes.
But this is just minor, as it will be clear in later sections.

marc-hb replied:
How about `manifest.projects.project.in_groups`. Besides avoiding the collision with the other key, `in_groups` also reduces IMHO the potential confusion between `- groupA` in this YAML list and `-groupA` elsewhere, I mean imagine someone completely new to west, then the `in_` prefix makes it IMHO harder to read too fast and mistake for a `-` exclusion:
```
  in_groups:
   - groupA
```


For the other key, how about: `manifest.groups_filter` as in Python list comprehension etc.

mbolivar-nordic replied:
> How about `manifest.projects.project.in_groups`

I'm not in favor of this change. At the API level, a `west.manifest.Project` has a `groups` attribute which corresponds to this data, so making this key `groups` is concise and consistent.

> For the other key, how about: manifest.groups_filter as in Python list comprehension etc.

I'm OK with something like this.

I guess you meant this to apply to both the configuration file and the manifest. This also impacts the `west update --groups` option.

I'd like to use `groups-filter` instead of `groups_filter` for consistency with other things (like `west-commands`, `repo-path`).

------------------------------------

**Summary of proposed change**

1. In YAML:

```yaml
manifest:
  projects:
     - name: myproject
       groups: # stays the same
         - foo
         - bar
  groups-filter: [-foo,-bar] # changes from 'groups' to 'groups-filter'
```

2. In config: the `manifest.groups` configuration option becomes `manifest.groups-filter`.
3. In command line options: the new `--groups` option in `west update` becomes `--groups-filter`.

------------------------------------

Thoughts from @marc-hb  @carlescufi @tejlmand and others?

marc-hb replied:
> I'm not in favor of this change.

OK, as long as one of them is renamed then they will differ from each other which I think is the main goal. That's why I tried to give options for either.

> I guess you meant this to apply to both the configuration file and the manifest. 

I didn't look that far but sure.

> I'd like to use groups-filter instead of groups_filter for consistency with other things (like west-commands, repo-path).

+1

tejlmand replied:
I think `groups` as now is fine, but I also think `groups-filter` is just as good.

No preference for one or the other.

carlescufi replied:
As long as we rename this to `group-filter`, I am OK with renaming the manifest-wide option, alongside the config and command-line ones.

tejlmand commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1183,357 @@ processed in this order:
+Active and Inactive Projects
+============================
+
+All projects are *active* by default. A project is *inactive* if it has groups,
+all of which are blocked. This is the only way to make a project inactive:
+projects with no groups are active, as are projects with one or more allowed
+groups.
```
If I didn't knew earlier discussion, this would be a very hard section to read.

How about:
```suggestion
All projects are *active* by default. Projects with no groups are always *active*.
A project is *inactive* if all the groups, to which it belongs, are blocked.
```

marc-hb replied:
This gave me pause too and I even tried to come with some better phrasing. Then I realize it takes some time to mentally picture some Venn diagrams from pure text and that my rephrasing attempts to shorten that time were making very little difference. I'm not convinced the suggestion is easier to parse.

This being said, have you thought about using the word "precedence", something like: "in other words, explicitly allowing takes precedence over blocking".

mbolivar-nordic replied:
@tejlmand 's suggestion has the benefit of being shorter. I'll take it.

tejlmand reviewed (changes requested):
Looking good.
Couple of minor comments.

marc-hb commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1183,357 @@ processed in this order:
+
+Group names are otherwise arbitrary strings. Group names are case sensitive.
+
+As a restriction, no project may use both ``import:`` and ``groups:``. (This
+avoids some edge cases whose semantics are difficult to specify.)
```
I disagree. Mentioning edge cases gives the important information this will likely never supported and some idea why in very few words. Dropping this can leave the wrong expectation that this could come in the near future.

mbolivar-nordic replied:
I think it's important for the reasons @marc-hb describes.

tejlmand replied:
accepted.

marc-hb commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,433 @@ processed in this order:
+.. code-block:: yaml
+
+   manifest:
+     groups: [-groupA,-groupB]
```
I think this syntax should support `[-groupA,+groupB,-groupC]` even if `+groupB` is a no-op. This would make it consistent with `manifest.groups` below, possibly make some scripting easier and also make the key name less strange because this `groups` is actually a list of `blocked_groups` in disguise for now.

marc-hb commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,433 @@ processed in this order:
+is a comma-separated list of groups. To allow a group, add its name to the
+list. To block a group, add its name prefixed by a dash (``-``). For example,
+setting ``manifest.groups`` to ``groupA,-groupB`` allows ``groupA``, and blocks
+``groupB``.
```
Unless it's too late but I think `+groupA,-groupB` would be more "consistent" than `groupA,-groupB` and it would also make a clearer difference from the `groups` homonym at the project level. It may even be easier to script config files changes if anyone is into that.

mbolivar-nordic replied:
> Unless it's too late but I think `+groupA,-groupB` would be more "consistent" than `groupA,-groupB` and it would also make a clearer difference from the `groups` homonym at the project level. It may even be easier to script config files changes if anyone is into that.

I would like @carlescufi and @tejlmand to weigh in here. It's not too late, but we should make a decision quickly if we can.

tejlmand replied:
> I would like @carlescufi and @tejlmand to weigh in here. It's not too late, but we should make a decision quickly if we can.

I like the proposal from @marc-hb.
Explicit having to do `+groupA` also remove the stupid user mistake of forgetting `-`, so he accidentially does `groupB` when he meant `-groupB`, `groupB` would throw an error at the user.
This greatly reduces one of those *doh* mistakes that can take time to track down.

mbolivar-nordic replied:
This alternate proposal requires us to forbid group names from starting with `+` or `-` -- are we OK with that?

mbolivar-nordic replied:
Sorry, to be more clear: they already can't start with `-`, but this would add `+` to the list of forbidden characters. I personally don't have issues with that, but want to get it out in front of people.

carlescufi replied:
Yes, I also agree with this proposal, it is more consistent now that we've settled on using this add/remove list of projects. Let's go with it.

tejlmand replied:
still agree to use `+groupA` and `-groupB` and thus restrict the use of `+` as first char of a group name.

marc-hb commented:
11 commits now, have you considered submitting the first and/or least debated commits separately to "divide and conquer"?

mbolivar-nordic commented:
> 11 commits now, have you considered submitting the first and/or least debated commits separately to "divide and conquer"?

I guess I could separate a few (three by my count) of these commits into their own PR, but then I would have two PRs to get merged, instead of one, and I'm not really sure that's worth the effort unless they're really inconveniencing you.

tejlmand commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,437 @@ processed in this order:
+             path: path/to/foo-second-sub
+       - name: bar
+
+Here, ``west update`` will initialize and update just the submodules in ``foo``
+with paths ``path/to/foo-first-sub`` and ``path/to/foo-second-sub``. Any
```
Based on this comment https://github.com/zephyrproject-rtos/zephyr/pull/31241/files#r557368457, I think a very important piece of information is needed here:
```suggestion
Here, ``west update`` will initialize and update just the submodules in ``foo``
with paths ``path/to/foo-first-sub`` and ``path/to/foo-second-sub``.  ``path/to/foo-first-sub`` and ``path/to/foo-second-sub`` will also have their submodules updated recursively. Any
```

tejlmand commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,437 @@ processed in this order:
+The ``submodules`` key may be a list of mappings. Each element in the list
+defines the name of the submodule to update, and the path -- relative to the
+project's absolute path in the workspace -- to use for the submodule.
+
```
Based on this comment https://github.com/zephyrproject-rtos/zephyr/pull/31241/files#r557368457, I think a very important piece of information is needed here.

```suggestion
The submodule will be updated recursively.

```

tejlmand reviewed (changes requested):
I think it is important to make users aware that any submodule being included will always be update recursively.

mbolivar-nordic commented:
Hi @carlescufi, @tejlmand, @marc-hb

I have now added the word "recursively" in four different places total in response to comments about submodule update. Please note also that the exact git commands used to update submodules were always given in repo-tool.rst, and they include the `--recursive` option. I hope we can consider the users sufficiently warned on this front now :smile: .

mbolivar-nordic commented:
I reviewed the conversations in this PR and here is an index into open questions:

- https://github.com/zephyrproject-rtos/zephyr/pull/31241#discussion_r557795465
- https://github.com/zephyrproject-rtos/zephyr/pull/31241#discussion_r557640860

Please provide feedback if you haven't already.

marc-hb commented:
> I guess I could separate a few (three by my count) of these commits into their own PR, but then I would have two PRs to get merged, instead of one, 

Reviewers save time with split PRs because it's easier to focus on fewer topics and review smaller lists of questions  in progress. I mean this would probably have not been needed with smaller PRs:

> I reviewed the conversations in this PR and here is _an index into open questions_

Especially problematic with github which is really not designed for reviewing series and does not offer anything anywhere close to `git range-diff`

>  and I'm not really sure that's worth the effort unless they're really inconveniencing you.

Focus is admittedly less of a problem for the submitter because they have to spend much more time on the task anyway.
In my experience however, even the submitter ends up "saving time" too because the reviewers are more productive and they are also more likely to find smaller gaps in their schedule to insert smaller context switches which means the whole process is shorter.

The "effort" does not require branches, shuffling commits with rebase -i and pushing HEAD~N is enough. This is especially quick with magit which I know you're using :-)

Now all this implies some commits are actually independent from other commits, when not then forget everything I just wrote; it does not apply. Also not worth it if you think we're nearly done (famous last words?)

A bit off-topic sorry. Well this PR is large anyway so it won't make much difference (see what I just did? :-D)

mbolivar-nordic committed:
doc: west: add project groups

This is a new feature that allows you to associate projects with
groups, then force west to ignore a subset of the projects list by
disabling project groups. This can be selectively overridden via
configuration file to make west stop ignoring the projects within an
individual workspace.

An example use case is to define a project whose Git repository
requires credentials to pull. Putting such a project in a group named
"foo" and disabling group "foo" would make "west update" ignore the
project unless the user specifically configures the workspace to allow
"foo". After that, "west update" will pull the project as usual.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
Latest force push addresses this comment thread: https://github.com/zephyrproject-rtos/zephyr/pull/31241#discussion_r561616456

tejlmand commented on doc/guides/west/release-notes.rst:
```diff
@@ -1,6 +1,70 @@
+  repository in addition to the project repository itself.
+
+- West manifests now support :ref:`west-manifest-groups`. Project groups can be
+  controlled via allowlists and blocklists to determine what projects are
```
Still needs an update here, from `allow` and `block`.

How about:
```suggestion
  controlled via group keys to determine what projects are
```

mbolivar-nordic replied:
Nice catch.

I think I'll go with:

```
- West manifests now support :ref:`west-manifest-groups`. Project groups can be
  enabled and disabled to determine what projects are "active", [...]
```

tejlmand commented on doc/guides/west/release-notes.rst:
```diff
@@ -1,6 +1,70 @@
+
+- ``west update`` no longer updates inactive projects by default. It now
+  supports a ``--groups`` option which allows for one-time modifications to the
+  project group allowlist / blocklist.
```
One more.
```suggestion
  project group's enable / disable settings.
```

mbolivar-nordic replied:
Would like to go with:

```
  supports a ``--groups`` option which allows for one-time modifications to
  the set of enabled and disabled project groups.
```

tejlmand reviewed (changes requested):
Found two occurrences where allowlist / blocklist were used with groups.

And gave it a try in describing the group key (feel free to improve further or discard).

ktaborowski commented on doc/guides/west/manifest.rst:
```diff
@@ -1179,6 +1189,450 @@ processed in this order:
+         url-base: https://git.example.com
+
+The ``manifest.groups`` configuration option is set to ``-groupA`` (you can
+ensure this by running ``west config manifest.groups -- -groupA``; the extra
```
It seems to me that `--` parameter is mandatory when disabling a group but obligatory when enabling it.
Can we unify group enable and disable commands in terms of adding `--` ?

mbolivar-nordic replied:
> It seems to me that `--` parameter is mandatory when disabling a group but obligatory when enabling it.

I don't understand why you say it is "obligatory" when enabling. Did you mean to say "optional"?

> Can we unify group enable and disable commands in terms of adding `--` ?

I'm not sure what you're asking for. The extra `--` is just a POSIX thing to make sure that the string "-groupA" in the command line is treated as a positional argument.

https://unix.stackexchange.com/questions/11376/what-does-double-dash-mean

You could write `west config manifest.groups -- +foo` and it would set `manifest.groups` to `+foo`, but it's not necessary to use `--` in this case because `+foo` is not ambiguous: it is definitely a positional argument because it does not begin with `-`.

ktaborowski replied:
What I was asking about using the `west config manifest.groups -- +foo` command for adding group and  `west config manifest.groups -- -foo` for removing. It just seems more coherent to me. But the actual description is also good.

mbolivar-nordic replied:
I see, thanks for explaining.

I try to avoid using `--` in `west config` lines in the documentation unless they are really necessary, because I don't want people to think that the ugly syntax is always required. The `--` is only needed when dealing with strings that start with `-`, so that is where I'd like to keep it. This is consistent with other places in the documentation, like here:

https://docs.zephyrproject.org/latest/guides/west/build-flash-debug.html#permanent-cmake-arguments

carlescufi reviewed (approved):
This is now in a very good state, thank you for all the effort an patience!

mbolivar-nordic commented:
@tejlmand @ktaborowski comments addressed, thank you for review.

mbolivar-nordic commented:
(I've also added some more internal forward references within the project groups section to make it easier to navigate and make it clearer from the beginning where the examples live.)

tejlmand reviewed (approved):
lgtm, thanks for the great effort.

mbolivar-nordic commented:
Last force push is just a rebase.

marc-hb reviewed (commented):
See `in_groups` and `groups_filter` suggestions below.

> groups can be enabled or disabled
> projects are active or inactive
> projects can be placed in an allowlist or blocklist when importing

LGTM ^^

mbolivar-nordic commented:
Latest push reflects updates here: https://github.com/zephyrproject-rtos/west/pull/470

mbolivar-nordic committed:
doc-build: build with west 0.9

This is required to get the relevant API docstrings.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
I've cut the release and updated the docs build workflow to install it from pypi. Removed DNM, should be good to go when CI is green.
